### DISCUSSION TITLE
Concerning the `parser-directories` setting and associated scanning

### DISCUSSION THREAD

**Discussion by sogaiu** [tree-sitter/tree-sitter#2049] [Ideas]
### Background

I think that there are at least two types of relatively distinct use cases of the tree-sitter cli:

* Working on a single language's grammar at a time
* Multiple language development and testing (e.g. injection, etc.)

My impression is that the `parser-directories` setting in `config.json` is geared more toward the latter use case.

IIUC, the setting allows one to specify _parent_ directories that are scanned for subdirectories with names beginning with `tree-sitter-` -- not precisely which directories should be considered.

### Problem

I believe this can lead to [unexpected results / confusion](https://github.com/tree-sitter/tree-sitter/issues/732) in combination with the [automatic compilation (and installation) feature](https://github.com/tree-sitter/tree-sitter/issues/2017) if one is not aware of how this functions.

In my own case, I sometimes work with multiple directories each of which has a differing version of a particular language repository.  Their directory names typically vary by a trailing prefix like `.feature-1` or `.feature-2` and I tend to keep most of the source I work on living under the same parent directory (e.g. ~/src).  I don't think this is an unusual setup though I haven't conducted a study :)

Unfortunately, AFAIU, in combination with the automatic compilation and scanning that the tree-sitter cli does under some circumstances https://github.com/tree-sitter/tree-sitter/issues/2017#issuecomment-1374932752, this can lead to multiple grammar repositories being picked up and I can end up with an unexpected `<languge>.so` file (say one for feature-1, instead of one for feature-2) in the location tree-sitter expects to find such things.  I initially spent quite some time investigating this after noticing that the `parse` command was producing results that seemed impossible.

I know now to use the `dump-languages` subcommand to get an idea of exactly which `tree-sitter-*` directories have been recognized but I didn't find this out until later.  Moreover, even knowing this I need to suspect something is amiss to investigate.  I think it would be better to have some way to choose to not have any other grammar repository directories considered if invoking `parse` (or other relevant subcommands) from within a particular grammar repository directory.

### Exploration

I've tried to think of a way to prevent this from happening that doesn't involve having to change how I store my source nor how I name the directories (I suspect other folks do similar things and not just for tree-sitter-related repositories).  I've also considered using separate machines, vms, other accounts, etc. but these seem like measures that one shouldn't have to resort to.

I did arrive at [a hack](https://github.com/tree-sitter/tree-sitter/issues/2017#issuecomment-1375796247) (see footnote via link) but think it would be better if there were a more straight-forward way to achieve the end (it's not a pretty hack).

One idea is to give meaning to an empty list / array value for the `parser-directories` setting in `config.json`.  The rough idea is to make that mean "only process the current directory (assuming it's a grammar repository directory)".  I tried out a rough version and described it [here](https://github.com/tree-sitter/tree-sitter/issues/2017#issuecomment-1396426999).  A downside of this method is that unless one arranges to have a `config.json` file per grammar repository (so relying on a single `config.json`), it's probably awkward to switch things off in the case where one wants to work with multiple languages (the other use case mentioned at the beginning of this post).

Another might be a command line option and/or environment variable to indicate the desired intent.

Perhaps there are better ideas though.

Any thoughts?

**Comment by leonerd** [comment]
I might be misreading or misunderstanding the problem, but this feels like the crux of the problem is that you have a bunch of "work-in-progress" branches or forks of parsers visible in directories listed in `config.json`. It kindof feels like the solution is "don't do that". I'd suggest only putting "production-ready" parsers in directories listed there, and any work-in-progress can happen elsewhere, away from the auto-discovery mechanism.
